cashews/wrapper/transaction.py

```diff
@@ -0,0 +1,410 @@
+        if self.current_tx:
+            await self.current_tx.commit()
+
+    async def rollback(self):
+        if self.current_tx:
+            await self.current_tx.rollback()
+
+
+class Transaction:
+    def __init__(self, mode: Optional[TransactionMode] = None, timeout: Optional[float] = None):
+        self._mode = mode
+        self._timeout = timeout
+        self._backends = {}
+
+    def wrap(self, backend: Backend) -> Backend:
+        return self._backends.setdefault(backend, self._get_tx_backend(backend))
+
+    def _get_tx_backend(self, backend: Backend) -> "TransactionBackend":
+        if self._mode == TransactionMode.FAST:
+            return TransactionBackend(backend)
+        if self._mode == TransactionMode.SERIALIZABLE:
+            return LockTransactionBackend(backend, serializable=True, timeout=self._timeout)
+        return LockTransactionBackend(backend, serializable=False, timeout=self._timeout)
+
+    async def commit(self):
+        for tx_backend in self._backends.values():
+            await tx_backend.commit()
+
+    async def rollback(self):
+        for tx_backend in self._backends.values():
+            await tx_backend.rollback()
+
+
+class TransactionBackend(Backend):
+    def __init__(self, backend: Backend):
+        self._backend = backend
+        self._local_cache = Memory()
+        self._to_delete = set()
+        super().__init__()
+
+    def _key_is_delete(self, key: str) -> bool:
+        if key in self._to_delete:
+            return True
+        return False
+
+    async def commit(self):
+        if self._to_delete:
+            await self._backend.delete_many(*self._to_delete)
+
+        expire_group = {}
+        for key, (expire, value) in self._local_cache.store.items():
+            if expire:
+                expire = int(expire - time.time())
+            expire_group.setdefault(expire, {})[key] = value
+
+        for expire, kv in expire_group.items():
+            await self._backend.set_many(kv, expire=expire)
+        self._clear_local_storage()
+
+    async def rollback(self):
+        self._clear_local_storage()
+
+    def _clear_local_storage(self):
+        self._local_cache = Memory()
+        self._to_delete = set()
+
+    async def set(
+        self,
+        key: str,
+        value: Any,
+        expire: Optional[float] = None,
+        exist: Optional[bool] = None,
+    ) -> bool:
+        if exist is not None:
+            if not await self._backend.exists(key) is exist:
+                return False
+        if key in self._to_delete:
+            self._to_delete.remove(key)
+        return await self._local_cache.set(key, value, expire, exist)
+
+    async def set_many(self, pairs: Mapping[str, Any], expire: Optional[float] = None):
+        for key in pairs.keys():
+            if key in self._to_delete:
+                self._to_delete.remove(key)
+        return await self._local_cache.set_many(pairs, expire)
+
+    async def incr(self, key: str) -> int:
```

Comment:
I have a question about expected behaviour of this method. Let's say I've made following changes:

```
async with cache.transaction(mode):
    cache.delete("key")
    # some other logic
    cache.incr("key")
```

Currently it will restore the key with its latest value and increment it, which a little contradicts the behaviour without transaction. Is it intentional?

Reply:
awesome - rewrote and added test